In [14]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn.feature_extraction.text import CountVectorizer

# 1. Menyiapkan 8 data contoh review produk sesuai tabel laporan
data_review = {
    "teks": [
        "Bagus banget, produk ori dan pengiriman cepat!",
        "Kecewa, barang rusak dan penjual slow respon.",
        "Kualitas mantap, sesuai deskripsi, suka sekali.",
        "Jelek, bahan tipis dan ukurannya kekecilan.",
        "Sangat puas, berfungsi dengan baik dan awet.",
        "Bintang satu, barang tidak sesuai foto, kapok.",
        "Respon penjual ramah, kurir cepat, top pokoknya.",
        "Parah, paket bocor dan isinya kurang."
    ],
    "label": [1, 0, 1, 0, 1, 0, 1, 0]  # 1 = Positif, 0 = Negatif
}

df = pd.DataFrame(data_review)

# 2. Preprocessing & Ekstraksi Fitur teks menjadi angka
vectorizer = CountVectorizer(lowercase=True)
X_vector = vectorizer.fit_transform(df["teks"])
y = df["label"]

# 3. Membangun Model Neural Network (Multi-Layer Perceptron)
mlp = MLPClassifier(hidden_layer_sizes=(4,), activation='logistic', solver='sgd', max_iter=500, random_state=42)

# 4. Proses Training Model (Forward Propagation, Error, & Backpropagation)
mlp.fit(X_vector, y)

# 5. Menguji Model dengan Teks Ulasan Baru di Luar Dataset
review_baru = ["Produknya bagus sekali, saya sangat puas dan pengiriman cepat"]
review_baru_vector = vectorizer.transform(review_baru)

# 6. Menampilkan Output Prediksi
prediksi = mlp.predict(review_baru_vector)
probabilitas = mlp.predict_proba(review_baru_vector)

print("=== SIMULASI OUTPUT NEURAL NETWORK ===")
print(f"Review Baru: '{review_baru[0]}'")
print(f"Probabilitas [Negatif, Positif]: {probabilitas[0]}")
print(f"Hasil Prediksi Kelas: {prediksi[0]} ({'Positif' if prediksi[0] == 1 else 'Negatif'})")

=== SIMULASI OUTPUT NEURAL NETWORK ===
Review Baru: 'Produknya bagus sekali, saya sangat puas dan pengiriman cepat'
Probabilitas [Negatif, Positif]: [0.5210085 0.4789915]
Hasil Prediksi Kelas: 0 (Negatif)


## Tinjauan Data
Berikut adalah 8 data contoh review produk yang digunakan dalam simulasi ini, dengan label `1` untuk ulasan positif dan `0` untuk ulasan negatif.

In [ ]:
display(df)

,teks,label
0,"Bagus banget, produk ori dan pengiriman cepat!",1
1,"Kecewa, barang rusak dan penjual slow respon.",0
2,"Kualitas mantap, sesuai deskripsi, suka sekali.",1
3,"Jelek, bahan tipis dan ukurannya kekecilan.",0
4,"Sangat puas, berfungsi dengan baik dan awet.",1
5,"Bintang satu, barang tidak sesuai foto, kapok.",0
6,"Respon penjual ramah, kurir cepat, top pokoknya.",1
7,"Parah, paket bocor dan isinya kurang.",0


## Penjelasan Fitur

Fitur yang digunakan dalam model ini adalah representasi vektor dari teks ulasan produk. Prosesnya adalah sebagai berikut:

1.  **CountVectorizer:** Kami menggunakan `CountVectorizer` dari `sklearn.feature_extraction.text`. `CountVectorizer` mengubah koleksi dokumen teks menjadi matriks jumlah token (kata) atau term frequency.
    *   `lowercase=True`: Semua teks dikonversi menjadi huruf kecil untuk memastikan bahwa kata yang sama dengan kapitalisasi berbeda (misalnya, 'Bagus' dan 'bagus') dihitung sebagai token yang sama.
    *   Setiap kolom dalam matriks `X_vector` merepresentasikan sebuah kata unik dari seluruh korpus teks ulasan, dan nilai di setiap sel adalah frekuensi kemunculan kata tersebut dalam ulasan tertentu.

Dengan demikian, setiap ulasan produk direpresentasikan sebagai sebuah vektor numerik, di mana setiap dimensi vektor sesuai dengan frekuensi kata tertentu. Ini memungkinkan model Neural Network untuk 'memahami' dan memproses informasi tekstual.

## Alur Belajar Model (Neural Network - MLPClassifier)

Model yang digunakan adalah `MLPClassifier` (Multi-Layer Perceptron), sebuah jenis *feedforward* Neural Network. Berikut adalah alur belajarnya:

1.  **Input (Masukan):**
    *   Input ke jaringan saraf adalah vektor fitur `X_vector` yang dihasilkan oleh `CountVectorizer`. Setiap vektor merepresentasikan sebuah ulasan produk.
    *   Setiap elemen dalam vektor input ini (yaitu, frekuensi kemunculan kata) dikirim ke lapisan input Neural Network.

2.  **Forward Propagation (Propagasi Maju):**
    *   Input dari lapisan input dikalikan dengan bobot (weights) yang terhubung ke neuron di lapisan tersembunyi (`hidden_layer_sizes=(4,)` berarti ada satu lapisan tersembunyi dengan 4 neuron).
    *   Hasil perkalian ini dijumlahkan, dan ditambahkan dengan bias.
    *   Jumlah ini kemudian dilewatkan melalui fungsi aktivasi (`activation='logistic'` atau sigmoid). Fungsi aktivasi ini memperkenalkan non-linearitas, memungkinkan jaringan untuk mempelajari pola yang kompleks.
    *   Output dari lapisan tersembunyi ini kemudian menjadi input untuk lapisan output.
    *   Di lapisan output, proses serupa terjadi (perkalian dengan bobot, penjumlahan dengan bias, dan fungsi aktivasi). Karena ini adalah masalah klasifikasi biner, lapisan output biasanya menghasilkan probabilitas untuk setiap kelas (misalnya, probabilitas positif dan negatif).

3.  **Output (Keluaran):**
    *   Output dari *forward propagation* adalah probabilitas prediksi untuk setiap kelas (Negatif atau Positif) untuk setiap ulasan. Misalnya, `[0.5210085, 0.4789915]` berarti ada probabilitas 52.1% review itu negatif dan 47.9% positif.
    *   Kelas prediksi akhir ditentukan berdasarkan probabilitas tertinggi (misalnya, jika probabilitas negatif > probabilitas positif, maka diprediksi 'Negatif').

4.  **Error (Kesalahan):**
    *   `MLPClassifier` membandingkan output prediksi dengan label sebenarnya (`y`).
    *   Perbedaan antara prediksi model dan label aktual dihitung sebagai *error* atau *loss*. Fungsi *loss* (misalnya, *cross-entropy loss* untuk klasifikasi) mengukur seberapa jauh prediksi model dari kebenaran.

5.  **Backpropagation (Propagasi Mundur):**
    *   Error yang dihitung pada langkah sebelumnya kemudian 'disebarkan' mundur melalui jaringan, dari lapisan output kembali ke lapisan input.
    *   Selama *backpropagation*, algoritma (`solver='sgd'` atau *Stochastic Gradient Descent*) menghitung *gradient* dari fungsi *loss* terhadap setiap bobot dan bias dalam jaringan. *Gradient* ini menunjukkan arah dan besarnya perubahan yang diperlukan pada bobot dan bias untuk mengurangi *error*.
    *   Bobot dan bias diperbarui secara iteratif menggunakan *gradient* dan *learning rate* (`max_iter=500` menentukan jumlah iterasi pelatihan). Tujuan utamanya adalah untuk meminimalkan fungsi *loss*.

Proses *forward propagation*, perhitungan *error*, dan *backpropagation* ini berulang sebanyak `max_iter` kali, atau hingga konvergensi, sehingga jaringan 'belajar' untuk membuat prediksi yang semakin akurat.

## Mengapa Neural Network Lebih Cocok daripada Model yang Terlalu Sederhana?

Untuk kasus klasifikasi teks seperti analisis sentimen ulasan produk, Neural Network (terutama yang lebih kompleks dari MLP sederhana ini) menawarkan beberapa keunggulan signifikan dibandingkan model yang *terlalu* sederhana (misalnya, *Naive Bayes* murni atau *Logistic Regression* tanpa *feature engineering* yang canggih).

1.  **Kemampuan Menangkap Pola Non-Linear:**
    *   Ulasan teks seringkali mengandung hubungan yang kompleks dan non-linear antar kata atau frasa yang berkontribusi pada sentimen. Kata-kata seperti 'tidak' atau 'bukan' dapat sepenuhnya mengubah sentimen sebuah kalimat. Model sederhana mungkin kesulitan menangkap nuansa ini.
    *   Neural Network, dengan adanya lapisan tersembunyi dan fungsi aktivasi non-linear (seperti `logistic` yang digunakan di sini), mampu mempelajari dan merepresentasikan pola-pola non-linear ini, sehingga dapat membuat keputusan klasifikasi yang lebih akurat dan canggih.

2.  **Pembelajaran Representasi Fitur Otomatis:**
    *   Pada model yang terlalu sederhana, *feature engineering* (misalnya, membuat fitur *n-gram* atau TF-IDF) seringkali merupakan langkah manual yang intensif dan membutuhkan pengetahuan domain. Meskipun `CountVectorizer` adalah bentuk *feature engineering* awal, Neural Network, terutama dengan lapisan yang lebih dalam, dapat belajar untuk mengekstrak fitur yang lebih abstrak dan relevan secara otomatis dari data input.
    *   Lapisan tersembunyi dalam NN dapat belajar 'representasi' internal dari input, misalnya, mengidentifikasi kelompok kata yang sering muncul bersamaan dengan sentimen tertentu, tanpa perlu kita secara eksplisit mendefinisikan fitur tersebut.

3.  **Penanganan Konteks dan Urutan (untuk NN yang lebih kompleks):**
    *   Meskipun `MLPClassifier` yang digunakan di sini tidak secara inheren menangani urutan kata, Neural Network yang lebih canggih seperti Recurrent Neural Network (RNN) atau Transformer secara khusus dirancang untuk memahami konteks dan urutan dalam data sekuensial seperti teks. Ini sangat penting untuk memahami makna sebenarnya dari sebuah kalimat atau ulasan, di mana urutan kata sangat mempengaruhi sentimen.
    *   Model sederhana tidak memiliki kemampuan ini dan memperlakukan kata-kata sebagai entitas independen.

4.  **Skalabilitas dan Kinerja dengan Data Besar:**
    *   Neural Network dapat berskala dengan baik ke dataset yang sangat besar dan, dengan sumber daya komputasi yang memadai (misalnya, GPU), dapat dilatih secara efisien.
    *   Dengan lebih banyak data, Neural Network memiliki potensi untuk belajar representasi yang lebih kaya dan membuat generalisasi yang lebih baik, mengungguli model sederhana yang mungkin mencapai batas kinerjanya lebih cepat.

5.  **Fleksibilitas:**
    *   Neural Network sangat fleksibel dan dapat diadaptasi untuk berbagai tugas NLP, tidak hanya klasifikasi sentimen, tetapi juga penerjemahan mesin, ringkasan teks, dan lainnya, dengan perubahan arsitektur yang relatif minimal.

Meskipun untuk dataset yang sangat kecil dan sederhana seperti contoh 8 ulasan ini, model sederhana mungkin memberikan hasil yang 'cukup', potensi penuh Neural Network bersinar ketika berhadapan dengan data teks yang lebih besar, lebih bervariasi, dan lebih kompleks, di mana kemampuan mereka untuk menangkap pola non-linear dan mempelajari representasi fitur secara otomatis menjadi sangat berharga.